src/examples/radiuss_tutorial/lesson_03/device_self_intersections.cpp

```diff
@@ -0,0 +1,550 @@
+  // Either way, bbox_v will be a view w/ data in the correct space
+  auto& bbox_h = triMesh.triangleBoundingBoxes();
+  BBoxArray bbox_d =
+    on_device ? BBoxArray(bbox_h, kernel_allocator) : BBoxArray();
+  auto bbox_v = on_device ? bbox_d.view() : bbox_h.view();
+
+  // copy indices of non-degenerate triangles into a new array
+  // we expect this to be fast enough to perform on the host
+  // valid_v will be a view to the array in the proper space
+  axom::utilities::Timer timer;
+  timer.start();
+  const auto totalTriangles = triMesh.numTriangles();
+  IndexArray valid_h(0, totalTriangles, host_allocator);
+  for(axom::IndexType idx = 0; idx < totalTriangles; ++idx)
+  {
+    if(!triMesh.isTriangleDegenerate(idx))
+    {
+      valid_h.push_back(idx);
+    }
+  }
+  auto valid_d = on_device ? IndexArray(valid_h, kernel_allocator) : IndexArray();
+  auto valid_v = on_device ? valid_d.view() : valid_h.view();
+
+  timer.stop();
+  SLIC_INFO_IF(
+    verboseOutput,
+    axom::fmt::format("Computing non-degenerate triangles took {:4.3} seconds.",
+                      timer.elapsedTimeInSec()));
+
+  // lambda to check if two triangles w/ given indices intersect
+  auto trianglesIntersect =
+    AXOM_LAMBDA(axom::IndexType idx1, axom::IndexType idx2)
+  {
+    constexpr bool includeBoundaries = false;  // only use triangle interiors
+
+    return (!useBoundingBoxes ||
+            axom::primal::intersect(bbox_v[idx1], bbox_v[idx2])) &&
+      axom::primal::intersect(tris_v[idx1], tris_v[idx2], includeBoundaries, tol);
+  };
+
+  // Phase I: Find the number of intersecting pairs of triangles
+  const auto validCount = valid_v.size();
+  RAJA::RangeSegment row_range(0, validCount);
+  RAJA::RangeSegment col_range(0, validCount);
+
+  using KERNEL_POL =
+    typename axom::mint::internal::structured_exec<ExecSpace>::loop2d_policy;
+  using REDUCE_POL = typename axom::execution_space<ExecSpace>::reduce_policy;
+  using ATOMIC_POL = typename axom::execution_space<ExecSpace>::atomic_policy;
+
+  RAJA::ReduceSum<REDUCE_POL, int> numIntersect(0);
+
+  // Compute the number of intersections
+  timer.start();
+  RAJA::kernel<KERNEL_POL>(
+    RAJA::make_tuple(col_range, row_range),
+    AXOM_LAMBDA(int col, int row) {
+      if(row < col && trianglesIntersect(valid_v[row], valid_v[col]))
+      {
+        numIntersect += 1;
+      }
+    });
+  timer.stop();
+  SLIC_INFO_IF(
+    verboseOutput,
+    axom::fmt::format("Finding intersection count took {:4.3} seconds.",
+                      timer.elapsedTimeInSec()));
+
+  // Phase II: If there are intersections, we need to extract their indices to an array
+  if(numIntersect.get() > 0)
+  {
+    const auto numIndices = 2 * numIntersect.get();
+
+    // allocate an array for the intersections (without initializing it)
+    IndexArray intersections_d(axom::ArrayOptions::Uninitialized {},
+                               numIndices,
+                               kernel_allocator);
```

Comment:
```suggestion
    IndexArray intersections_d(axom::ArrayOptions::Uninitialized {},
                               numIndices,
                               numIndices,
                               kernel_allocator);
```

Bugfix for CUDA policy. Lesson 4 also has a few instances of this constructor call as well. Looks like the uninitialized constructor does work, just uses a different overload. @kennyweiss reported this works on rzansel as-is, maybe it's the Automatic Translatation Service covering for us?

Reply:
Thanks @bmhan12

My solution was to use the initializing constructor -- e,g,
```cpp
    IndexArray intersections_d(numIndices,numIndices,kernel_allocator);
```
Yours is better!

I'll update the code for `lesson_03` and `lesson_04`.